src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs

```diff
@@ -166,137 +252,165 @@ private async Task MonitorServerAsync()
+                        averageRoundTripTime: averageRoundTripTimeRounded,
+                        canonicalEndPoint: heartbeatIsMasterResult.Me,
+                        electionId: heartbeatIsMasterResult.ElectionId,
+                        lastWriteTimestamp: heartbeatIsMasterResult.LastWriteTimestamp,
+                        logicalSessionTimeout: heartbeatIsMasterResult.LogicalSessionTimeout,
+                        maxBatchCount: heartbeatIsMasterResult.MaxBatchCount,
+                        maxDocumentSize: heartbeatIsMasterResult.MaxDocumentSize,
+                        maxMessageSize: heartbeatIsMasterResult.MaxMessageSize,
+                        replicaSetConfig: heartbeatIsMasterResult.GetReplicaSetConfig(),
+                        state: ServerState.Connected,
+                        tags: heartbeatIsMasterResult.Tags,
+                        type: heartbeatIsMasterResult.ServerType,
+                        version: _handshakeBuildInfoResult.ServerVersion,
+                        wireVersionRange: new Range<int>(heartbeatIsMasterResult.MinWireVersion, heartbeatIsMasterResult.MaxWireVersion));
+                }
+                else
+                {
+                    newDescription = _baseDescription.With(lastUpdateTimestamp: DateTime.UtcNow);
+                }
+
+                if (heartbeatException != null)
+                {
+                    if (heartbeatException is MongoCommandException heartbeatCommandException)
+                    {
+                        var topologyVersion = TopologyVersion.FromMongoCommandException(heartbeatCommandException);
+                        newDescription = newDescription.With(heartbeatException: heartbeatException, topologyVersion: topologyVersion);
+                    }
+                    else
+                    {
+                        newDescription = newDescription.With(heartbeatException: heartbeatException);
+                    }
+                }
+
+                newDescription = newDescription.With(reasonChanged: "Heartbeat", lastHeartbeatTimestamp: DateTime.UtcNow);
+
+                SetDescription(newDescription);
+
+                immediateAttempt =
+                    // serverSupportsStreaming
+                    (newDescription.Type != ServerType.Unknown && heartbeatIsMasterResult != null && heartbeatIsMasterResult.TopologyVersion != null) ||
+                    // connectionIsStreaming
+                    (isMasterProtocol != null && isMasterProtocol.MoreToCome) ||
+                    // transitionedWithNetworkError
+                    (IsNetworkError(heartbeatException) && previousDescription.Type != ServerType.Unknown);
             }
 
-            if (heartbeatException != null)
+            bool IsCancellationException(Exception ex, bool currentCheckCancelled, CancellationToken token)
             {
-                var topologyVersion = default(Optional<TopologyVersion>);
```

Comment:
Style comment: I believe @rstam was the one who suggested this particular refactor in my  CSHARP-3057 PR, so I'll defer to Robert as whether he wants this particular style restored :)